### PR TITLE
Grand Cleanup & Stuff

### PR DESCRIPTION
**Pull request by kobolog**
A lot of tiny annoying crap:
  – Fixed a possible dangling reference in `context_t::locate()` (and broke Elliptics plugin – TBD).
  – Made `signal<T>` more predictable in when it triggers signals.
  – Fixed Locator's streams not being closed when a client with the same UUID asks for a new stream.
  – Dropped `boost::karma` for comma-sperated string concatenation in favor of `boost::join()`.
  – Moved almost all synchronized blocks to `locked_ptr<T>::apply()` calls for better traces.
  – Added more verbosity in log & error messages.
  – Flattened and simplified code all around the place.
  – Cleaned up Zipkin code a little bit.

NOTE: Commit squashing & rebasing TBD.

### CODE REVIEW

File: include/cocaine/api/cluster.hpp [hunk 1]
```diff
@@ -1,6 +1,6 @@
 /*
-    Copyright (c) 2011-2014 Andrey Sibiryov <me@kobology.ru>
-    Copyright (c) 2011-2014 Other contributors as noted in the AUTHORS file.
+    Copyright (c) 2011-2015 Andrey Sibiryov <me@kobology.ru>
```
Comment (3Hren) [inline include/cocaine/api/cluster.hpp]: Out of the scope. Please send separate PR with these changes.

File: include/cocaine/format.hpp [hunk 2]
```diff
@@ -43,11 +43,11 @@ substitute(boost::format&& message, const T& argument, const Args&... args) {
 template<class... Args>
 static inline
 std::string
-format(const std::string& format, const Args&... args) {
+format(const std::string& format, Args&&... args) {
```
Comment (3Hren) [inline include/cocaine/format.hpp]: Strictly speaking there is no need for universal references here. Moreover it can hurt.

Given for example you have a large struct with overloaded `operator<<` - `foo_t`, and `sizeof(foo_t) == 1000`. Then when we type `format(foo_t())` the argument is deduced as an rvalue.

Thus on every `forward` we "move" those 1000 bytes by copying, which is worse than just passing a const reference.

File: include/cocaine/rpc/actor.hpp [hunk 3]
```diff
@@ -26,8 +26,18 @@
 
 #include <asio/ip/tcp.hpp>
 
+#include "cocaine/rpc/graph.hpp"
+
 namespace cocaine {
 
+class quote_t
+{
```
Comment (3Hren) [inline include/cocaine/rpc/actor.hpp]: Style?

File: include/cocaine/rpc/dispatch.hpp [hunk 4]
```diff
@@ -62,8 +64,21 @@ class basic_dispatch_t {
...
     virtual
-    boost::optional<dispatch_ptr_t>
-    process(const decoder_t::message_type& message, const upstream_ptr_t& upstream) const = 0;
+    auto
+    process(int id, const rpc_t& rpc) const -> boost::optional<dispatch_ptr_t> = 0;
+
+    // Observers
+
+    auto
+    name() const -> std::string;
```
Comment (3Hren) [inline include/cocaine/rpc/dispatch.hpp]: Declarations like `auto fn(args_t...) -> result_type` looks ugly and here there is no need for them.

File: include/cocaine/rpc/dispatch.hpp [hunk 5]
```diff
@@ -99,11 +140,12 @@ class dispatch:
 
     // Slot construction
 
+    typedef mpl::lambda
+        <std::shared_ptr<io::basic_slot<mpl::_1>>> slot_ptr;
```
Comment (3Hren) [inline include/cocaine/rpc/dispatch.hpp]: Again, the new era of code style?

File: include/cocaine/rpc/graph.hpp [hunk 6]
```diff
@@ -25,6 +25,8 @@
 #include <string>
 #include <tuple>
 
+#include <boost/functional/hash_fwd.hpp>
```
Comment (3Hren) [inline include/cocaine/rpc/graph.hpp]: Why not `std::hash`?

File: include/cocaine/rpc/graph.hpp [hunk 7]
```diff
@@ -52,6 +54,15 @@ typedef std::map<
     std::tuple<std::string, boost::optional<graph_node_t>, boost::optional<graph_node_t>>
 > graph_root_t;
 
+// Support for computing hashes of protocols. Unless two different protocols have identical message
+// names in exactly the same spots in the graph, with exactly the same transitions, it'll be unique.
```
Comment (3Hren) [inline include/cocaine/rpc/graph.hpp]: They'll be **probably** unique.

File: include/cocaine/rpc/slot/deferred.hpp [hunk 8]
```diff
@@ -118,12 +118,6 @@ struct deferred {
         return *this;
     }
 
-    deferred&
```
Comment (3Hren) [inline include/cocaine/rpc/slot/deferred.hpp]: Random code moving, please send another PR with style.

File: src/context.cpp [hunk 9]
```diff
@@ -155,34 +147,35 @@ context_t::remove(const std::string& name) {
...
-    }
+        if (it != list.end()) {
+            auto  endpoints = it->second->endpoints();
+            auto& prototype = it->second->prototype();
 
-    return boost::optional<const actor_t&>(it->second->is_active(), *it->second);
-}
+            // TODO(@kobolog): Figure out if there should always be some endpoints for a
+            // service. Useless is_active() check was dropped, so adding this assert JIC.
+            BOOST_ASSERT(!endpoints.empty());
```
Comment (3Hren) [inline src/context.cpp]: Asserting on possible results is strange. Why'd just not throw an exception?

File: src/errors.cpp [hunk 10]
```diff
@@ -102,13 +96,11 @@ class dispatch_category_t:
         if(code == cocaine::error::dispatch_errors::uncaught_error)
             return "uncaught invocation exception";
 
-        return "cocaine.rpc.dispatch error";
+        return cocaine::format("generic %s error", name());
```
Comment (3Hren) [inline src/errors.cpp]: C++ does it more verbosely: `::format("%s %d", name(), code);` which result in `cocaine.rpc.dispatch: 42`.

File: src/errors.cpp [hunk 11]
```diff
@@ -150,7 +140,28 @@ class security_category_t:
...
+    virtual
+    auto
+    name() const throw() -> const char* {
+        return "cocaine.service.locator";
+    }
+
+    virtual
+    auto
+    message(int code) const -> std::string {
+        if(code == cocaine::error::locator_errors::service_not_available)
```
Comment (3Hren) [inline src/errors.cpp]: Why the fuck there are so much branches? Do we need them? Even more - without curly braces around the success statement.

The `switch` case looks better here.

File: src/service/locator.cpp [hunk 12]
```diff
@@ -118,9 +93,35 @@ class locator_t::connect_sink_t: public dispatch<event_traits<locator::connect>:
...
+    typedef io::protocol<event_traits<locator::connect>::upstream_type>::scope protocol;
+
+    on<protocol::chunk>(std::bind(&connect_sink_t::on_announce, this, ph::_1, ph::_2));
+    on<protocol::choke>(std::bind(&connect_sink_t::on_shutdown, this));
+}
+
+locator_t::connect_sink_t::~connect_sink_t() {
+    auto lock = parent->m_clients.synchronize();
+
+    for(auto it = active.begin(); it != active.end(); ++it) tuple::invoke(
```
Comment (3Hren) [inline src/service/locator.cpp]: Why not range-based for?

File: src/service/locator.cpp [hunk 13]
```diff
@@ -118,9 +93,35 @@ class locator_t::connect_sink_t: public dispatch<event_traits<locator::connect>:
...
+    });
+
+    cleanup();
+}
+
 void
 locator_t::connect_sink_t::discard(const std::error_code& ec) const {
-    if(ec.value() == 0) return;
+    if(ec.value() == 0) {
+        return;
```
Comment (3Hren) [inline src/service/locator.cpp]: :+1:

File: src/service/locator.cpp [hunk 14]
```diff
@@ -205,121 +196,141 @@ locator_t::connect_sink_t::on_shutdown() {
...
+private:
+    void
+    on_discard(const std::error_code& ec, const std::string& handle);
+};
 
-            std::tie(versions, protocol) = metadata;
+auto
+locator_t::publish_slot_t::operator()(tuple_type&& args, upstream_type&& upstream) -> result_type {
+    const auto dispatch = cocaine::tuple::invoke(std::move(args),
+        [this](std::string&& handle, std::vector<tcp::endpoint>&& location,
```
Comment (3Hren) [inline src/service/locator.cpp]: Forced rvalue references here looks... strange. We can just pass by value, aren't we?

File: src/service/locator.cpp [hunk 15]
```diff
@@ -669,38 +680,46 @@ locator_t::on_routing(const std::string& ruid, bool replace) -> streamed<results
...
+        const auto update = results::connect{m_cfg.uuid, {{name, meta}}};
 
-    if(mode == modes::exposed) {
-        if(m_snapshots.count(name) != 0) {
-            COCAINE_LOG_ERROR(m_log, "duplicate service detected");
-            return;
-        }
+        if(mode == modes::exposed) {
+            if(m_snapshots.count(name) != 0) {
+                COCAINE_LOG_ERROR(m_log, "duplicate service detected");
```
Comment (3Hren) [inline src/service/locator.cpp]: I think it'll be great if this message would contain the duplicated service name.

File: src/service/locator.cpp [hunk 16]
```diff
@@ -714,20 +733,20 @@ locator_t::on_context_shutdown() {
...
-        m_cluster = nullptr;
-
         if(mapping.empty()) {
             return;
         } else {
-            COCAINE_LOG_DEBUG(m_log, "closing %d outgoing locator streams", mapping.size());
+            COCAINE_LOG_DEBUG(m_log, "closing %d outgoing locator stream(s)", mapping.size());
         }
 
         boost::for_each(mapping | boost::adaptors::map_values, [](streamed<results::connect>& s) {
```
Comment (3Hren) [inline src/service/locator.cpp]: Range based for!!!

File: src/service/locator.cpp [hunk 17]
```diff
@@ -714,20 +733,20 @@ locator_t::on_context_shutdown() {
...
-            try { s.close(); } catch(...) { /* None */ }
+            try { s.close(); } catch(...) { /***/ }
```
Comment (3Hren) [inline src/service/locator.cpp]: I think we should never silently eat exceptions.

File: src/session.cpp [hunk 18]
```diff
@@ -199,25 +206,28 @@ session_t::handle(const decoder_t::message_type& message) {
...
             )});
 
             max_channel_id = channel_id;
         }
-        if(lb->second->upstream->client_trace) {
-            incoming_trace = lb->second->upstream->client_trace;
-        } else {
+
+        if(!lb->second->upstream->trace) {
             auto trace_header = message.meta<hpack::headers::trace_id<>>();
```
Comment (3Hren) [inline src/session.cpp]: Looks ugly (here and several lines below) for my opinion.

**Comment by 3Hren** [comment]
PRs are for granulated features. Instead of squashing I suggest to split this PR.

File: include/cocaine/rpc/graph.hpp (hunk 6 above)
Comment (kobolog) [inline include/cocaine/rpc/graph.hpp]: It doesn't support hashing of maps.

File: include/cocaine/api/cluster.hpp (hunk 1 above)
Comment (kobolog) [inline include/cocaine/api/cluster.hpp]: Don't be ridiculous — it's not even a code change.

Comment (3Hren) [inline include/cocaine/api/cluster.hpp]: It's a diff.

We use licenses with ancient rules, which requires special header for each file - okay. But it's change is a separate logic unit.

File: src/context.cpp (hunk 9 above)
Comment (kobolog) [inline src/context.cpp]: It shouldn't be empty in all cases I THINK, hence assert. I'm going to refactor actor_t, so I keep it here to assert the expected variant.

File: include/cocaine/rpc/graph.hpp (hunk 6 above)
Comment (3Hren) [inline include/cocaine/rpc/graph.hpp]: I haven't known, okay.

File: src/errors.cpp (hunk 10 above)
Comment (kobolog) [inline src/errors.cpp]: Won't your `error::to_string` bullshit do it for us?

Comment (3Hren) [inline src/errors.cpp]: It does, but for known errors. For unknown errors it shows nothing, because you eat those information.

File: src/errors.cpp (hunk 11 above)
Comment (kobolog) [inline src/errors.cpp]: Because it's already like that above, so here I'm simply making it look the same.

Comment (3Hren) [inline src/errors.cpp]: Okay, but it looks inappropriately for int switching.

File: src/service/locator.cpp (hunk 12 above)
Comment (kobolog) [inline src/service/locator.cpp]: Because 1) we never used it before 2) no code is changed here.

File: src/service/locator.cpp (hunk 14 above)
Comment (kobolog) [inline src/service/locator.cpp]: Again, no code changed here.

Comment (kobolog) [inline src/service/locator.cpp]: Again, no code changed here – but possible good point.

File: src/service/locator.cpp (hunk 15 above)
Comment (kobolog) [inline src/service/locator.cpp]: It will be in the "service" attribute, obviously.

File: src/service/locator.cpp (hunk 16 above)
Comment (kobolog) [inline src/service/locator.cpp]: No code change, never used it around codebase.

File: src/service/locator.cpp (hunk 15 above)
Comment (3Hren) [inline src/service/locator.cpp]: Okay, I forgot about it.

File: src/service/locator.cpp (hunk 17 above)
Comment (kobolog) [inline src/service/locator.cpp]: What do you want to happen here then? Spam logs with "we tried to close a stream on core shutdown but it was already closed"?

Comment (3Hren) [inline src/service/locator.cpp]: Yes, only the truth.

File: src/session.cpp (hunk 18 above)
Comment (kobolog) [inline src/session.cpp]: Speak to the Braille Guy, that's his creation.

Comment (3Hren) [inline src/session.cpp]: Okay =)

File: include/cocaine/rpc/dispatch.hpp (hunk 4 above)
Comment (kobolog) [inline include/cocaine/rpc/dispatch.hpp]: I think they look amazingly beautiful.

Comment (3Hren) [inline include/cocaine/rpc/dispatch.hpp]: Actually no.

File: include/cocaine/rpc/actor.hpp (hunk 3 above)
Comment (kobolog) [inline include/cocaine/rpc/actor.hpp]: I used this style in other places where the first line of the class definition body is one of the access specifiers.

File: include/cocaine/rpc/dispatch.hpp (hunk 5 above)
Comment (kobolog) [inline include/cocaine/rpc/dispatch.hpp]: Typo!

File: include/cocaine/rpc/actor.hpp (hunk 3 above)
Comment (3Hren) [inline include/cocaine/rpc/actor.hpp]: Okay then we have mixed style. Which of them more "correct"?.

File: src/session.cpp (hunk 18 above)
Comment (3Hren) [inline src/session.cpp]: @antmat look! It's your creation!

File: src/errors.cpp (hunk 10 above)
Comment (kobolog) [inline src/errors.cpp]: It will be `[${ec.value()}] ${what()}: ${ec.message()}` (at least with libc++), what are you talking about?

Comment (3Hren) [inline src/errors.cpp]: Right, my fault.

File: include/cocaine/format.hpp (hunk 2 above)
Comment (kobolog) [inline include/cocaine/format.hpp]: Good point, I'll revert.

File: include/cocaine/rpc/actor.hpp (hunk 3 above)
Comment (kobolog) [inline include/cocaine/rpc/actor.hpp]: More "correct" is probably to convert this one into a `struct`.

File: include/cocaine/rpc/dispatch.hpp (hunk 4 above)
Comment (kobolog) [inline include/cocaine/rpc/dispatch.hpp]: I think that 'name (input_types_tuple) -> (output_types_tuple)' is how all function declarations should look in all PLs because it's streamlined and makes sense. This is as close as it gets in C++.

File: src/session.cpp (hunk 18 above)
Comment (3Hren) [inline src/session.cpp]: Actually only variable names looks ugly for me here.

File: src/service/locator.cpp (hunk 17 above)
Comment (kobolog) [inline src/service/locator.cpp]: I think in this case we can and should silently ignore exception since it is the shutdown path, i.e. not only nobody cares about it, but also we can't really do anything about errors here.

**Comment by 3Hren** [comment]
Any progress on this issue?
